torch/distributed/distributed_c10d.py

```diff
@@ -902,35 +955,42 @@ def _new_process_group_helper(
     is_default_group = len(global_ranks_in_group) == 0
 
-    backend = Backend(backend)
-    pg: Union[ProcessGroupGloo, ProcessGroupMPI, ProcessGroupNCCL, ProcessGroupUCC]
-    if backend == Backend.MPI:
-        if not is_mpi_available():
-            raise RuntimeError(
-                "Distributed package doesn't have MPI built in."
-                " MPI is only included if you build PyTorch from"
-                " source on a host that has MPI installed."
-            )
-        pg = ProcessGroupMPI.create(global_ranks_in_group)
-        if not pg:
+    # If this is a subgroup (which means group_ranks is specified),
+    # we check if the current process is a member of the new group.
+    if not is_default_group:
+        global_rank = _get_default_group().rank()
+        if global_rank not in global_ranks_in_group:
             return GroupMember.NON_GROUP_MEMBER
-    else:
-        # If this is a subgroup (which means group_ranks is specified),
-        # we check if the current process is a member of the new group.
-        if not is_default_group:
-            global_rank = _get_default_group().rank()
-            if global_rank not in global_ranks_in_group:
-                return GroupMember.NON_GROUP_MEMBER
 
+    prefix_store = PrefixStore(f"{group_name}/", store)
+    base_pg_options = ProcessGroup.Options(backend=str(backend))
+    base_pg_options._timeout = timeout
+    pg: ProcessGroup = ProcessGroup(prefix_store, group_rank, group_size, base_pg_options)
+
+    for device, backend_str in backend_config.get_device_backend_map().items():
         # Use the group name as prefix in the default store, such that
         # a single store can be reused by multiple groups.
-        prefix_store = PrefixStore(group_name, store)
+        prefix_store = PrefixStore(f"{device}/", prefix_store)
 
-        if backend == Backend.GLOO:
-            if pg_options is not None:
-                raise RuntimeError("GLOO options not supported")
-            pg = ProcessGroupGloo(prefix_store, group_rank, group_size, timeout=timeout)
-        elif backend == Backend.NCCL:
+        if backend_str == Backend.MPI:
+            if not is_mpi_available():
+                raise RuntimeError(
+                    "Distributed package doesn't have MPI built in."
+                    " MPI is only included if you build PyTorch from"
+                    " source on a host that has MPI installed."
+                )
+            backend_pg = ProcessGroupMPI.create(global_ranks_in_group)
+            backend_pg_type = ProcessGroup.BackendType.MPI
+            if not backend_pg:
+                return GroupMember.NON_GROUP_MEMBER
+
+        if backend_str == Backend.GLOO:
+            # TODO: remove this check after lazy initialization is supported
+            # if pg_options is not None:
+            #     raise RuntimeError("GLOO options not supported")
+            backend_pg = ProcessGroupGloo(prefix_store, group_rank, group_size, timeout=timeout)
```

Comment:
Do we need to create Backend at Python level? Can we create Backend under PG at cpp level? (You already have code for that in ProcessGroup.cpp)